HW2/P5/P5.txt

```diff
@@ -0,0 +1,137 @@
+
+Problem 2. Grid
+----------------
+Results
+~~~~~~~
+1 Thread
+683.9007777 simulation frames per second
+651.76787026 simulation frames per second
+674.598900715 simulation frames per second
+639.163027818 simulation frames per second
+580.542575573 simulation frames per second
+724.980813773 simulation frames per second
+707.064059339 simulation frames per second
+726.719292531 simulation frames per second
+655.546234726 simulation frames per second
+695.981581222 simulation frames per second
+680.811816764 simulation frames per second
+566.890572426 simulation frames per second
+
+4 Threads
+1309.86412287 simulation frames per second
+1315.05369969 simulation frames per second
+1224.0477187 simulation frames per second
+1073.40185837 simulation frames per second
+1047.48362422 simulation frames per second
+885.646555365 simulation frames per second
+1079.32939234 simulation frames per second
+1051.8630721 simulation frames per second
+563.354729998 simulation frames per second
+1150.65211978 simulation frames per second
+1076.89689258 simulation frames per second
+1034.55774259 simulation frames per second
+
+Analysis
+~~~~~~~~
+Introducing the grid yields some incredible results. On one thread, the speedup is approximately 100x to a range of 550-730 and on 4 threads the speedup is a bit less, around 80x from the 4-threaded gridless version. This speedup is expected since the algorithm is linear instead of polynomial. At the same time, on the small test simulation, it seems to be that the balls overlap more frequently and to a greater degree than they did without the grid. Though I cannot confirm this in any systematic manner, it may be the case that we gain a performance boost by trading off collision detection accuracy. Ultimately though, the combination of algorithm optimization and multithreading boosts performance very noticeably (though the algorithm improvement clearly holds more weight).
+
+Problem 3. Sorting
+----------------
+Results
+~~~~~~~
+1 Thread
+701.579275305 simulation frames per second
+903.706084321 simulation frames per second
+872.601386963 simulation frames per second
+872.262644188 simulation frames per second
+905.827770463 simulation frames per second
+871.061022464 simulation frames per second
+869.702240844 simulation frames per second
+900.501585982 simulation frames per second
+873.503967345 simulation frames per second
+779.92822354 simulation frames per second
+
+4 Threads
+1644.55540268 simulation frames per second
+2649.96809757 simulation frames per second
+2404.22495382 simulation frames per second
+1709.7589452 simulation frames per second
+2251.73668633 simulation frames per second
+2512.88486941 simulation frames per second
+2345.35929614 simulation frames per second
+1996.52136351 simulation frames per second
+1998.60866102 simulation frames per second
```

Comment:
Wow!  My z-order sorting did not give me results anywhere near this good.  This is really impressive.